ERCS/erc-7920.md

```diff
@@ -0,0 +1,358 @@
+      ["0x4bdbac3830d492ac3f4b0ef674786940fb33481b32392e88edafd45d507429f2"],
+      ["0x95be87f8abefcddc8116061a06b18906f32298a4644882d06baff852164858c6"]
+    ]
+  }
+}
+```
+
+## Rationale
+
+### UX Improvement
+
+A single signature that covers multiple messages
+
+### Isolated Verification
+
+Independent verification of messages without knowledge of others
+
+### Human-readable
+
+This ERC preserves the readability benefits of EIP-712. Giving wallets and users insight into what is being signed.
+
+### Efficient verification on-chain
+
+`_verifyMerkleProof` has a runtime of `O(log2(N))` where N is the number of messages that were signed.
+
+When N=1, merkleRoot **must** equal keccak256(encode(m₁)) and proofs MUST be empty arrays
+
+### Improved Wallet Security
+
+Certain messages signed in isolation may appear harmless but combined with may be harmful. Giving the full list of messages to users could help them better navigate their experience.
+
+### Flexible Verification Modes
+
+Applications can require combination of messages be signed together to enhance security.
+
+## Backwards Compatibility
+
+When the number of message is one, `eth_signTypedData_v5` produces the same signature as `eth_signTypedData_v4` since `merkleRoot == keccak256(encode(message))`. This allows `eth_signTypedData_v5` to be a drop-in replacement for `eth_signTypedData_v4` with no changes to on-chain verification.
+
+## Reference Implementation
+
+Reference implementation of `eth_signTypedData_v5` can be found the assets directory.
+
+## Security Considerations
+
+### Replay Protection
+
+This ERC focuses on generating composite messages and verifying their signatures. It does not contain mechanisms to prevent replays. Developers **must** ensure their applications can handle receiving the same message twice.
+
+### Partial Message Verification
+
+During verification, care **must** be taken to ensure that **both** of these checks pass:
+
+1. EIP-712 signature on the Merkle root is valid
+2. Merkle proof is valid against the root
+
+### User Understanding
+
+Wallets **must** communicate to users that they are signing multiple messages at once. Wallets **must** display of all message types before signing.
+
+To ensure batch signature requests are digestible, it is recommended max number of messages is at most 10.
+
+### Merkle Tree Construction
+
+Merkle tree should be constructed in a consistent manner.
+
+1. The hashing function **must** be `keccak256`
+2. To ensure predictable/consistent proof sizes, implementations **must** pad leaves with zero hashes to reach next power of two to ensure balance. Let `n` be the number of messages. Before constructing the tree, compute the smallest `k` such that `2^(k-1) < n ≤ 2^k`. Insert zero hashes into the list of messages until list of messages is equal to `2^k`.
+3. To ensure an implicit verification path, pairs **must** be sorted lexicographically before constructing parent hash.
+
+#### Tree Reference Implementation
```

Comment:
Large code snippets like this should be in the Reference Implementation section. You can use an `[anchor link](#some-heading)` to refer to it.